File: Algorithm/328. Odd Even Linked List.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-06-19 14:08:01
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-19 14:08:17
# @Email: liangchaowu5@gmail.com

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        count, odd, even, t1, t2= 1, None, None, None, None
        while head:
            if count % 2 == 1:
                if odd == None: 
                    odd = head
                    t1 = head
                else:
                    t1.next = head
                    t1 = head
            else:
                if even == None: 
                    even = head
                    t2 = head
                else:
                    t2.next = head
                    t2 = head
            head = head.next
            count += 1
        if t1 != None:
            t1.next = even
        if t2 != None:
            t2.next = None
        return odd

File: Algorithm/Java/328. Odd Even Linked List.java
/*
* @Author: LC
* @Date:   2017-08-28 23:56:28
* @Last Modified by:   LC
* @Last Modified time: 2017-08-28 23:56:48
*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// two pointers
class Solution 
{
    public ListNode oddEvenList(ListNode head) 
    {
        if (head == null || head.next == null) return head;
        ListNode p1 = head, p2 = head.next, head2 = head.next;
        while(p1.next != null && p2.next != null)
        {
            p1.next = p2.next;
            p1 = p1.next;
            p2.next = p1.next;
            p2 = p2.next;
        }
        p1.next = head2;
        return head;
    }
}